### PR TITLE
[BFT Testing] Adds integration test for ingress corruption

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR fills in the missing parts of enabling ingress traffic, and develops an integration test involving all the components of the BFT testing framework. The integration test is running on a mock engine A with a corruptible conduit and another honest engine B where both engines have their own separate networking layer (over stub network).
An orchestrator network is established and connects to the corrupted networking layer that engine A resides on.
Honest engine B sends a message to corrupted engine A, i.e., an ingress message. The message should go through the orchestrator network to a mock orchestrator that completely overrides its content with some fixture, and is routed back to engine A to be processed. 
The test is passed if engine A receives the corrupted message that the orchestrator composed on behalf of honest engine B.

### CODE REVIEW

File: insecure/orchestrator/network.go [hunk 1]
```diff
@@ -107,28 +107,46 @@ func (on *Network) stop() error {
...
-// Instead of dispatching their messages to the networking layer of Flow, the conduits of corrupt nodes
-// dispatch the outgoing messages to the orchestrator network by calling the InboundHandler method of it remotely.
+// Instead of dispatching their messages to the networking layer of Flow, the corrupt nodes
+// dispatch both the incoming (i.e., ingress) as well as the outgoing (i.e., egress)
+// messages to the orchestrator network by calling the InboundHandler method of it remotely.
 func (on *Network) Observe(message *insecure.Message) {
-	if err := on.processEgressMessage(message); err != nil {
-		on.logger.Fatal().Err(err).Msg("could not process message of corrupt node")
+	if message.Ingress != nil && message.Egress != nil {
+		// In BFT testing framework, it is a bug to has both ingress and egress set.
```
Comment (gomisha) [inline insecure/orchestrator/network.go:115]:
```suggestion
		// In BFT testing framework, it is a bug to have both ingress and egress messages not set.
```

File: insecure/orchestrator/network.go [hunk 2]
```diff
@@ -107,28 +107,46 @@ func (on *Network) stop() error {
...
 // Observe is the inbound message handler of the orchestrator network.
-// Instead of dispatching their messages to the networking layer of Flow, the conduits of corrupt nodes
-// dispatch the outgoing messages to the orchestrator network by calling the InboundHandler method of it remotely.
+// Instead of dispatching their messages to the networking layer of Flow, the corrupt nodes
+// dispatch both the incoming (i.e., ingress) as well as the outgoing (i.e., egress)
+// messages to the orchestrator network by calling the InboundHandler method of it remotely.
 func (on *Network) Observe(message *insecure.Message) {
-	if err := on.processEgressMessage(message); err != nil {
-		on.logger.Fatal().Err(err).Msg("could not process message of corrupt node")
+	if message.Ingress != nil && message.Egress != nil {
```
Comment (gomisha) [inline insecure/orchestrator/network.go:114]: Do we need to add the 2nd error condition of having both ingress and egress messages set?

**Review by peterargue — APPROVED** [review]

File: insecure/orchestrator/network.go (hunk 2 above)
Comment (yhassanzadeh13) [reply insecure/orchestrator/network.go:114]: Yes, per our conversation, exchanging an `insecure.Message` with both the ingress and egress set is an illegal state.

File: insecure/corruptible/network.go [hunk 3]
```diff
@@ -110,13 +116,23 @@ func NewCorruptNetwork(
 // Except, it first wraps the given processor around a corrupt message processor, and then
 // registers the corrupt message processor to the original flow network.
 func (n *Network) Register(channel channels.Channel, messageProcessor flownet.MessageProcessor) (flownet.Conduit, error) {
-	corruptProcessor := NewCorruptMessageProcessor(n.logger, messageProcessor)
+	n.mu.Lock()
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply insecure/corruptible/network.go]: Per our conversation refactored to concurrent map: https://github.com/onflow/flow-go/pull/3215/commits/252b185c3a46d8392c4a52ece07a98c1f2f5e884

File: insecure/corruptible/network.go [hunk 4]
```diff
@@ -48,6 +48,11 @@ type Network struct {
 	conduitFactory        insecure.CorruptibleConduitFactory
 	attackerInboundStream insecure.CorruptibleConduitFactory_ConnectAttackerServer // inbound stream to attack orchestrator
 
+	// message processors for each channel.
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply insecure/corruptible/network.go]: https://github.com/onflow/flow-go/pull/3215/commits/189a5741932f7aadfc455ebcbb0ccc35ad49063d

File: insecure/corruptible/network.go [hunk 5]
```diff
@@ -184,6 +203,65 @@ func (n *Network) ProcessAttackerMessage(stream insecure.CorruptibleConduitFacto
...
+		Hex("corrupted_target_id", logging.ID(targetId)).
+		Str("flow_protocol_event_type", fmt.Sprintf("%T", event)).Logger()
+
+	if targetId != n.me.NodeID() {
+		lg.Fatal().Msg("corrupt network received ingress message for a different node")
+	}
+
+	lg.Info().Msg("corrupt network received ingress message")
+
+	originalProcessor, err := n.originalMessageProcessorForChannel(channels.Channel(msg.ChannelID))
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply insecure/corruptible/network.go]: https://github.com/onflow/flow-go/pull/3215/commits/d80d6d486ff1fe11bc65b22ebae2708918715393

**Pull request merged**